### ISSUE TITLE
Deprecate OnFailure(), move all usages to MapFailure

### ISSUE THREAD

**Issue by vkhorikov** [vkhorikov/CSharpFunctionalExtensions#148]
Similar to the recent refactoring in #141, I propose to:

1. Refactor `OnFailure` into `MapFailure` and mark `OnFailure` as `[Obsolete]`. This will bring extension methods to consistency with F#'s `Result` at https://github.com/fsharp/fsharp/blob/master/src/fsharp/FSharp.Core/result.fs (`MapFailure` and not `MapError` because `Error` and `IsFailure` are different members in our `Result` and `Failure` is closer semantically to describing the result's state).

2. Move the existing `MapFailure` methods out of `Result`/`Result<T>`/`Result<T,E>` to extensions that replace `OnFailure` ( https://github.com/vkhorikov/CSharpFunctionalExtensions/blob/master/CSharpFunctionalExtensions/Result/Result.cs#L554-L567 and similar).

@space-alien let me know if you can do that. If you are short on time, I'll try to refactor myself this weekend.

**Comment by space-alien** [comment]
Suggestion (2) may not be ideal, as an extension method would complicate this usage:

```csharp
Result<T> result = Result.Ok<T>();
resultT.MapFailure<NewType>();
```

An extension method would not be able to infer T type param, so user would have to write:
`resultT.MapFailure<OldType, NewType>();` which is not ideal.

We could do this:
`resultT.MapFailure(typeof(NewType))` - still a bit ugly.

I'll look into #1 in the meantime.

**Comment by vkhorikov** [comment]
Will #151 help with (2) ? If not, I'm good with leaving it there. (1) is more important.

**Comment by space-alien** [comment]
I've looked into this.

Our current `OnFailure` overloads are actually "on failure, `Tap`".

Meanwhile, `OnFailureCompensate` is currently a mix of "on failure, `Map`" and "on failure, `Bind`", where we have overloads that risk discarding `E` in the process (similar to those I have just removed from `Map` that disregarded `T`).

Finally, `MapFailure`, doesn't translate to "on failure, `Map`". It's "on failure, `Map`, else throw".

We need to think about this some more.

**Comment by space-alien** [comment]
I remembered [your comment](https://github.com/vkhorikov/CSharpFunctionalExtensions/issues/135#issuecomment-527274792) on #135. I had also been thinking along these lines for a while.

One option would be to refactor as follows:

```csharp
Map(Func<K> onSuccess, Func<K> onFailure)
MapSuccess(Func<K> func) // i.e. the current Map behaviour (if starting with a failure, return a failure result of the new type)
MapFailure(Func<K> func) // Opposite of MapSuccess
```

And the same principle for `Bind()` and `Tap()`:
```csharp
Bind(Func<Result> onSuccess, Func<Result> onFailure)
BindSuccess(Func<Result> func)
BindFailure(Func<Result> func)

Tap(Action onSuccess, Action onFailure)
TapSuccess(Action action)
TapFailure(Action action)
```

Fortunately, I don't think any of this would be a breaking change (even with the current `MapFailure`).

I'm still trying to decide if I have an issue with the semantics of `MapFailure` / `MapSuccess`, and their Bind/Tap equivalents. For the reader unfamiliar with the library, it may not be instantly obvious whether:
1. The func is mapped/tapped/bound IF the starting result is in the named success/failure state, or
2. The func is ALWAYS called and the returned result will be changed into the named success/failure state.

It may be pretty daft to infer (2), but any ambiguity potentially draws some of the reader's brainpower to reach the right conclusion. And one of the benefits of this library has always been its ability to make expected behaviour obvious.

So, here's the "blindingly-obvious" alternative..!

```csharp
Map()
Tap()
Bind()

OnSuccessMap()
OnSuccessTap()
OnSuccessBind()

OnFailureMap()
OnFailureBind()
OnFailureTap()
```

**Comment by vkhorikov** [comment]
Gave some thought to all this. I agree with your analysis. A couple points:

- Regarding naming -- how about `MapOnSuccess()` and `MapOnFailure()` instead of `OnSuccessMap()` and `OnFailureMap()`?

- With these additions, looks like we don't need `Match` and `Finally` anymore, they are replaced by ``Tap``'s and ``Map``'s.

- The current `MapFailure` doesn't follow the functional programming paradigm as it's useful in imperative scenarios only (more specifically, in scenarios with `return` statements). It's still useful, though, because there are cases where plain old `return` is needed. But because its current name conflicts with the new additions, I suggest we rename it into `ConvertFailure`. No need for backward compatibility -- it's still a new feature. /cc @altmann

**Comment by space-alien** [comment]
@vkhorikov 
> Gave some thought to all this. I agree with your analysis. A couple points:
> 
> * Regarding naming -- how about `MapOnSuccess()` and `MapOnFailure()` instead of `OnSuccessMap()` and `OnFailureMap()`?

Yes, for some reason, I settled on the latter, but now I seem to be changing my mind again! I think it was because of the potential fluency in reading: `result.OnSuccessMap(ToSomething)`. But I don't feel strongly either way. `MapOnSuccess` has the advantage of bringing the actual behaviour to the front of the word.

> * With these additions, looks like we don't need `Match` and `Finally` anymore, they are replaced by `Tap`'s and `Map`'s.

Also my first thought, but then I remembered those methods yield a value instead of a Result, so I think they could still be useful at the end of the chain. We could always review the names if necessary.

**Comment by vkhorikov** [comment]
I was thinking about making the two-parameter `Map()` return the underlying `T` value too, otherwise there's not much point in it accepting two functions because we tell `Map()` what to do with the result in both situations -- when it's a success and when it's a failure.

So, `Map()` should be either (1)

```
K Map<K>(Func<K> onSuccess, Func<K> onFailure)
```

or (2)

```
Result<K> Map<K>(Func<K> onSuccess)
```

And now I think I'm changing my mind too. (1) is actually a `Match`, and should stay as such. `Map` shouldn't really leave the realm of the `Result` elevated space. Hence, I suggest the following signatures (using the most generic `Result` -- `Result<T,E>`):

```
// Changes T1 to T2
Result<T2, E> Map(this Result<T1, E> result, Func<T1, T2> onSuccess)

// No need for MapOnSuccess

// Changes E1 to E2
Result<T, E2> MapOnFailure(this Result<T, E1> result, Func<E1, E2> onFailure)
```

Same with `Bind` and `Tap`:

```
Tap()
TapOnFailure()

Bind()
BindOnFailure()
```

And have `Match()` deal with the two params situations:

```
R Match<T, E, R>(this Result<T, E> result, Func<T, R> onSuccess, Func<E, R> onFailure)
void Match<T, E>(this Result<T, E> result, Action<T> onSuccess, Action<E> onFailure)
// etc
```

**Comment by renato04** [comment]
@space-alien  There are a lot of  `OnFailure` methods in `OnFailure.cs`. Will all of them marked as deprecated? And suggest using `ConvertFailure`?

**Comment by vkhorikov** [comment]
@renato04 `ConvertFailure` is just to change the type of the `Result`, it's not related to the `OnFailure` extension methods. `OnFailure` needs to be renamed (while preserving backward compatibility) into the corresponding `MapFailure`, `BindFailure` and so on, similarly how we did it with `OnSuccess` methods.
